### PR TITLE
fix 190 is the max the index takes with a utf8mb4 table

### PR DESCRIPTION
**Pull request by fvbock**
on mysql

### CODE REVIEW

**Comment by bodhi** [comment]
I think this scheme of using composite natural primary keys is fundamentally flawed if MySQL enforces a limit on the length of the index.

Sure, you can say, "but keys/locales shouldn't be that long", but that's not really the point.

Is there a *disadvantage* to using a surrogate key? Actually, what is the point of having locale+key be the primary key to begin with? Is this because Gorm requires some kind of primary key? Or are you using it as a substitute for a unique index?

**Comment by fvbock** [comment]
i think the main point is that the gorm struct tax syntax currently does not allow to define field and index independently - something that would translate to 

```
CREATE TABLE `translations` (
    `key` varchar(255),
    `locale` varchar(255),
    `value` varchar(255) , 
    PRIMARY KEY (
        `key`(100),
        `locale`(100)
    )
);
```

or a unique version of that.

**Comment by bodhi** [comment]
> i think the main point is that the gorm struct tag syntax currently does not allow to define field and index independently

Is that going to be hard to fix? @jinzhu?

**Comment by jinzhu** [comment]
@bodhi no, gorm doesn't require that. the reason we are using multiple primary key is this could avoid duplicated translations, that has occurred for our Qor2 projects already.

So I want to using the unique index to fix that. (haven't thought about the limitation of index before ;( )

Do you have any better solution to do that? or just remove the unique index? that should works well.

**Comment by jinzhu** [comment]
> i think the main point is that the gorm struct tag syntax currently does not allow to define field and index independently - something that would translate to

@fvbock gorm does support that, will it resolve the problem if we do that?

**Comment by bodhi** [comment]
> just remove the unique index? that should works well.

Assuming you mean validating the uniqueness in the application instead? That will open up race conditions that will allow duplicate keys. The way to prevent it is to have a unique index in the database.

---

> we are using multiple primary key is this could avoid duplicated translations

Right. But there is no reason for this to be a *primary* index, it just needs to be unique.

That said, its is kind-of moot anyway, as:

> By default, an index key for a single-column index can be up to **767 bytes**. The same length limit applies to any index key prefix. See Section 13.1.13, “CREATE INDEX Syntax”. For example, you might hit this limit with a column prefix index of more than 255 characters on a TEXT or VARCHAR column, assuming a UTF-8 character set and the maximum of 3 bytes for each character. When the innodb_large_prefix configuration option is enabled, this length limit is raised to 3072 bytes, for InnoDB tables that use the DYNAMIC and COMPRESSED row formats.

So you can't do both of these on MySQL :

* Support arbitrary-length keys
* Ensure uniqueness of keys per locale

(Also, seems that the max length on Postgres is 8191).

File: i18n/backends/database/database.go [hunk 1]
```diff
@@ -26,8 +36,18 @@ func (backend *Backend) LoadTranslations() []*i18n.Translation {
 	return translations
 }
 
-func (backend *Backend) SaveTranslation(t *i18n.Translation) {
+func (backend *Backend) SaveTranslation(t *i18n.Translation) (err error) {
+	if utf8.RuneCountInString(t.Key) > 190 {
```
Comment (bodhi) [inline i18n/backends/database/database.go]: This (and `20`, on R44 below), shouldn't be duplicated with the struct definition? Is it possible in Go to define the value "190" is a constant that is also used in the struct tag?

**Pull request closed without merging**